README
unlogo
By Jeff Crouse
jefftimesten at gmail dot com
http://www.jeffcrouse.info
 
This software is licensed under the GNU-GPL version 3 or later.

unlogo is free software; you can redistribute it and/or
modify it under the terms of the GNU General Public
License as published by the Free Software Foundation; either
version 3 of the License, or (at your option) any later version.

unlogo is distributed in the hope that it will be useful,
but WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
General Public License for more details.



INTRODUCTION
==============================================
Unlogo is a web service that eliminates logos and other corporate signage from videos. On a practical level, it takes back your personal media from the corporations and advertisers. On a technical level, it is a really cool combination of some brand new OpenCV and FFMPEG functionality. On a poetic level, it is a tool for focusing on what is important in the record of your life rather than the ubiquitous messages that advertisers want you to focus on.

In short, Unlogo gives people the opportunity to opt out of having corporate messages permanently imprinted into the photographic record of their lives.


GOALS
==============================================

For this inital version of Unlogo, I am not interested in performance, only accuracy.  This has lead to several decisions, not the least of which is the fact that it is a video processing plugin rather than an AR experience or something of that nature.  


REQUIREMENTS
==============================================

Unlogo requires Boost (filesystem), FFMPEG (compiled with vf_plugin.c) and OpenCV.  I've provided installation instructions that have worked for me, but this is no guarantee that they will work for you!


FFMPEG compiled with vf_plugin.c
----------------------
Prerequisites

You may want to use your favorite package manager to install the following packages before compiling FFMPEG.  (links provided for manual install)

- lame 		http://downloads.sourceforge.net/project/lame/lame/3.98.4/lame-3.98.4.tar.gz
- x264 			ftp://ftp.videolan.org/pub/videolan/x264/snapshots/x264-snapshot-20100922-2245.tar.bz2
- libogg 			http://downloads.xiph.org/releases/ogg/libogg-1.2.0.tar.gz
- vorbis 		http://downloads.xiph.org/releases/vorbis/libvorbis-1.3.1.tar.gz
- xvid 			http://downloads.xvid.org/downloads/xvidcore-1.2.2.tar.gz
- faac 			http://downloads.sourceforge.net/faac/faac-1.28.tar.gz (had to use ./configure --disable-dependency-tracking --without-mp4v2) 
- theora 		http://downloads.xiph.org/releases/theora/libtheora-1.1.1.tar.bz2
- opencore-amr  http://downloads.sourceforge.net/project/opencore-amr/opencore-amr/0.1.2/opencore-amr-0.1.2.tar.gz

There are some excellent instructions here on installing FFMPEG prerequisites on Ubuntu: https://wiki.ubuntu.com/ffmpeg


Compiling & Installing

vf_plugin.c is a simple framework that allows you to specify a dylib or .o file to use as an avfilter in FFMPEG.  It basically saves you the trouble of having to recompile FFMPEG every time you make a change to your filter.  Perhaps future versions fo the moustachizer/Unlogo will just be a self-contained avfilter, but for development, this is the way to go*.

1. I start with the v0.6.1 tag from the git repository

git clone git://git.videolan.org/ffmpeg.git ffmpeg
cd ffmpeg
git checkout v0.6

2. add vf_plugin.c to the libavfilter folder

cp [this folder]/libavfilter/vf_plugin.c [your ffmpeg clone]/libavfilter

3. modify '[your ffmpeg clone]/libavfilter/allfilters.c' and [your ffmpeg clone]/libavfilter/Makefile to include vf_plugin (see examples in [this folder]/libavfilter)

NOTE:  Do not just copy the Makefile and allfilters.c that came with Unlogo.  

Makefile: add at the end of the list of lines that look the same
OBJS-$(CONFIG_PLUGIN_FILTER)                 += vf_plugin.o

allfilters.c: add at the end of the list of lines that look the same
REGISTER_FILTER (PLUGIN,	  plugin,	   vf);

4. ./configure

NOTE:  these are my configure options.  You may find them useful if you installed all of the prerequisites listed above.

./configure --arch=x86_64 --enable-libmp3lame --enable-shared  \
	--disable-mmx --disable-asm --disable-stripping --enable-debug  \
	--disable-doc --enable-libx264 --enable-nonfree --enable-libvorbis --enable-gpl --enable-libxvid --enable-version3 \
	--enable-pthreads --enable-libfaac --enable-libopencore-amrnb \
	--enable-libopencore-amrwb --enable-libtheora--enable-x11grab \
	--enable-ffplay --enable-avfilter

5. If you get an error, you probably just need to adjust your configure options.  You can run ./configure --help to see all of the options.  

If you didn't get any errors, 'make && sudo make install'



* The frei0r video filter framework is similar (but much more sophisticated) to vf_plugin.c, but I couldn't get it to work at all.


OpenCV 2.2 - http://opencv.willowgarage.com/wiki/
----------------------
The following process worked for me:

	wget http://downloads.sourceforge.net/project/opencvlibrary/opencv-unix/2.2/OpenCV-2.2.0.tar.bz2
	tar -xvf OpenCV-2.2.0.tar.bz2
	cd OpenCV-2.2.0
	ccmake -G "Unix Makefiles" 

At this point, you are in the 'ccmake' program, whose job it is to generate your Makefile.  Press 'c' to configure, and you should see all of the configuration options.  Change them as you like, then press 'g'.  This will generate the makefile and exit out of ccmake.  Then just:

make && sudo make install



EXAMPLE USAGE
==============================================
Unlogo uses logo databases generated by 'logomunge'.  Please see the logomunge utility in this distribution.


ffmpeg -i movies/mymovie.mp4 -qscale 2 -vf plugin=unlogo.dylib:logos/Adidas/variant1.xml.gz:logos/Adidas/variant2.xml.gz -y out.mov


SUPPORTERS
==============================================
I'd like to thank all of the people below for their contributions to the Unlogo Kickstarter campaign.

futnuh, Rod Parker, Daniel Doyle, Nathan Loofbourrow, Daniel Alavi, Martha Chase, Michael Lewis, Michael Schieben, con.sume.org, Steev, Hise, Michael Bridgen, Nick Semenkovich, Shawn Allen, Darren Nelsen, Jason Fitch, gabriel vogt, Graham Reznick, Michal Migurski, Ed Costello, Siddiq Bello, ayah, Geary Feldstein, Daniel, Klainbaum, max raven, Steve Dietz, Cary Aurand, Travis Thatcher, kiran, Kib T., J.R. Maroney, Vasili Sviridov, Gladys Santiago, Alexey Zilber, Curt Sherman, Meredith Binder, doug jaeger, Sean Parker, craigp, Barbara Iverson, Blake Eggemeyer, Parker Higgins, Philipp Schlösser, WhiteNitro, Alastair Medford, giorgio, Justyna B. // aconite, David Proctor, jodi morrison, dmandzio, Ken Nickerson, CJ Cornell, Jeremy Levine , R. Lee Montgomery, Owocni.pl, Dominic Es, Aaron Meyers, Adam T, Michael Mandiberg, Pseudo, Kristofer Bergstrom, Philip, uniomni, Jeff Ubois, Starchy Grant, Steve Lambert, Jon Cohrs, Matthew Katz, Zach, David Schmidt, Cassandra Gero, nortd, Eric, Val Gough, Ben Tomassetti, Toby Marvin, brekk, Melissa Cohen, Drew Fister, Craig Caryl, Louai Abu-Osba, William Linn, David Papini, ddkk, Volker Neumann, Caitlin Rothermel, Sam Fleischner, J. Makary, David Miller, Kendall Lister, Damien Bargiacchi, ViKing, Ryan Anthony Petersen, Simone Cortesi, Hans Czajkowski Jørgensen, Andre Theisen, Aaron Cossey, Ivan Wong, Robert Prus, Noah Keating, Carol Thomas, Matt Campen, Sunny Bates, Wesley Ross, DBonar, Ryan Kelln, Stephanie Pereira, Michelle, Christoph Richter, Brook Hinton, Gal Buki, james nelson, Krister Knutars, Jason Asbahr, John De Boer, LigoranoReese, Hypothetical Development Organization, Ted Timmons, Paul Dembina, Melissa 'Bunni' Elian, Don Burke, Hatim Belyamani, Rufus Deuchler, Kristian Bjørnard, Jason Evers, Paul Hamingson, laudowicz, kaho, Brett Camper



TO DO
==============================================

1. All matchers should implement hasHomography(), getHomography(), and warpPoints(), bInited, bTrainImageSet
2. Make the matcher classes more consistent in general -- perhaps make an interface that they must conform to
3. Make the train data save names consistent



CHANGELOG
CHANGELOG
---------------
I am starting this log less for changes, and more to keep track of the research that I am doing.  

Before I start, the state of Unlogo is currently as follows:

1.  The avfilter interface for FFMPEG allows developers to write their own filters that users can apply to videos when running ffmpeg, ffplay, etc.   There isn't very good documentation for it yet, but there are some tutorials available:  

http://wiki.multimedia.cx/index.php?title=FFmpeg_filter_howto
http://www.ffmpeg.org/doxygen/trunk/structAVFilter.html

Essentially, it asks developers to create a AVFilter avfilter_vf_plugin structure and add it to the libavfilter module before compiling FFMPEG.  Recompiling FFMPEG is kind of a pain when you are actively developing a filter, so instead, we have vf_plugin.c, which is a filter that takes, as an argument, a shared object or shared library with 3 functions exposed: "init", "uninit", and "process". I have compiled this into FFMPEG and provided instructions in README

So, to use the "plugin" avfilter, you would call ffmpeg like:

ffmpeg -i yourmovie.mov -vf plugin=[yourplugin.dylib]:list:of:additional:arguments  outfile.mpv

2. src/main.cpp contains the 3 required functions mentioned above.  It also contains an "unlogo" object.  init() simply passes the argument string (see :list:of:additional:arguments above) to unlogo::init().  uninit() simply deletes the instance of the unlogo object.  process() is where all of the interesting things happen.  NOTE:  It might seem like having this main.cpp is unnecessary, but I am very concerned with readabiltiy and encapsulation as much as possible.  I'd like to be able to have someone look at the unlogo class and not worry about the inner workings of vf_plugin.c or any of the FFMPEG stuff, so I have opted for this organization.

3.  in main.cpp, the process() function takes a bunch of arguments from FFMPEG.  These contain the pixels of a single frame. I initialize a cv::Mat object with the relevant pixels (src_stride[0] -- see cv::Mat constructor #8 in Mat.rtf), and then pass it to the unlogo object to be processed. You will notice that I also save a reference to the address of the pixel data and then check to make sure it hasn't changed after being passed to "process". This is because the cv::Mat object will sometimes re-allocate its uchar* data; if you try an operation that requires more memory. The next step is to copy those src pixels to the dst pixels, so it would be bad if we were operating on pixels that we thought were going to be copied into the output movie, but the cv::Mat had in fact reallocated them.

4. Moving into the unlogo class: At this level of abstraction, we have the Image class, which is a wrapper around cv::Mat.  The first thing we do is point an Image called query to the Mat passed in by main.cpp (Image::useMat).  So there are essentially 2 wrappers around the raw pixels provided by FFMPEG:  cv::Mat and then Image.  The actual pixels provided by FFMPEG are now located at query.cvImg.data. I chose not to extend cv::Mat because of problems passing the superclass to OpenCV functions.  This was mostly out of lack of c++ skill, but I think it's slightly more readable also.

5.  At this point I adopt the Train/Query relationship.  In step 2 I mentioned that main.cpp calls unlogo::init.  unlogo::init opens files (logos) that are specified by the ffmpeg command call.  These are the "train" images, or the ones that we are searching for.  Each frame that enters the process function is a "query".  

6.  There are also a bunch of different Matcher classes that we "train" with the logos and then run each frame.  They provide a Homography that can then be used to warp and translate a replacement graphic over the query image if the train is found.


This is basically the structure that I am working with right now. After several false starts where the code quickly became overwhelmingly complex, I think this is as simple and straightforward as it can be. Although I have tools for matching, I haven't done enough tests to know exactly how to implemnt them, and that is what I am working on now. 






March 6 2011 - Feature Vector Databases
===============================================

After seeing the letter_recog.cpp sample in opencv/samples/cpp, I spent today researching  how to use datasets on the UCI Machine Learning site.  In particular, this one: http://archive.ics.uci.edu/ml/datasets/Letter+Recognition

I was hoping that I would be able to use this as an additional layer of matching on top of the Matchers that I have been making. It has 20000 "feature vectors" of letter data. I was not familliar with the terminology or basic ideas of this kind of matching, so it took me a while to understand how you would even go about using the provided "letter-recognition.data" file.  As it turns out, each dataset is made up of one or more "classes" (in this case, each letter represents a class).  Each class may have 1 or more entries in the dataset.  The first column of a row is a label of a class, which is followed by several numbers that define an instance of the class.  After some digging, I found that the numbers in letter-recognition.data correspond to (on a scale of 0-15):

 1.	lettr	capital letter	(26 values from A to Z)
 2.	x-box	horizontal position of box	(integer)
 3.	y-box	vertical position of box	(integer)
 4.	width	width of box			(integer)
 5.	high 	height of box			(integer)
 6.	onpix	total # on pixels		(integer)
 7.	x-bar	mean x of on pixels in box	(integer)
 8.	y-bar	mean y of on pixels in box	(integer)
 9.	x2bar	mean x variance			(integer)
10.	y2bar	mean y variance			(integer)
11.	xybar	mean x y correlation		(integer)
12.	x2ybr	mean of x * x * y		(integer)
13.	xy2br	mean of x * y * y		(integer)
14.	x-ege	mean edge count left to right	(integer)
15.	xegvy	correlation of x-ege with y	(integer)
16.	y-ege	mean edge count bottom to top	(integer)
17.	yegvx	correlation of y-ege with x	(integer)


install.sh
#!/bin/bash

set -e
set -o pipefail

SDK=/Developer/SDKs/MacOSX10.6.sdk
CFLAGS="-msse -arch x86_64 -arch i686 -nostdinc -g -pipe  \
    -B$SDK/usr/include/gcc -B$SDK/usr/lib/gcc \
    -isystem$SDK/usr/include -F$SDK/System/Library/Frameworks"
LDFLAGS="-arch x86_64 -arch i686 -Wl,-syslibroot,$SDK "
CXXFLAGS="$CFLAGS"
export SDK CFLAGS CXXFLAGS LDFLAGS


WGET=wget
type -P wget &>/dev/null || { 
WGET=curl -L -O
}

clear
echo -------------------------
echo "Welcome to the FFMPEG installer"
echo "CFLAGS $CFLAGS"
echo "CPPFLAGS $CFLAGS"
echo "LDFLAGS $LDFLAGS"
echo -------------------------


read -p "Press any key to start install"
echo "Making the distribution directory..."




set -e
set -o pipefail
SCRIPT=$(cd ${0%/*} && echo $PWD/${0##*/})
PREFIX=`dirname $SCRIPT`
mkdir -p $PREFIX/dist


# Use this method to see if each library is installed
# http://serverfault.com/questions/54736/how-to-check-if-a-library-is-installed




if [ -e "$PREFIX/lib/libmp3lame.a" ]
then
	echo "LAME is already installed"
else
	clear
	echo -------------------------
	echo "Downloading and building LAME"
	echo -------------------------
	cd $PREFIX/dist
	${WGET} http://downloads.sourceforge.net/project/lame/lame/3.98.4/lame-3.98.4.tar.gz
	tar -xvf lame-3.98.4.tar.gz
	cd lame-3.98.4
	./configure --prefix=$PREFIX -enable-shared=no
	make && make install
fi



if [ -e "$PREFIX/lib/libyasm.a" ]
then
	echo "YASM is already installed"
else
	clear
	echo -------------------------
	echo "Downloading and building YASM"
	echo -------------------------
	cd $PREFIX/dist
	${WGET} http://www.tortall.net/projects/yasm/releases/yasm-1.1.0.tar.gz
	tar -xvf yasm-1.1.0.tar.gz
	cd yasm-1.1.0
	./configure --prefix=$PREFIX
	make && make install
fi



if [ -e "$PREFIX/lib/libfaac.a" ]
then
	echo "FAAC is already installed"
else
	clear
	echo -------------------------
	echo "Downloading and building FAAC"
	echo -------------------------
	cd $PREFIX/dist
	${WGET} http://downloads.sourceforge.net/faac/faac-1.28.tar.gz
	tar -xvf faac-1.28.tar.gz
	cd faac-1.28
	./configure --prefix=$PREFIX --enable-shared=no
	make && make install
fi



if [ -e "$PREFIX/lib/libopencore-amrnb.a" ]
then
	echo "opencore-arm is already installed"
else
	clear
	echo -------------------------
	echo "Downloading and building opencore-arm"
	echo -------------------------
	cd $PREFIX/dist
	${WGET} http://downloads.sourceforge.net/project/opencore-amr/opencore-amr/0.1.2/opencore-amr-0.1.2.tar.gz
	tar -xvf opencore-amr-0.1.2.tar.gz
	cd opencore-amr-0.1.2
	./configure --enable-shared=no --prefix=$PREFIX
	make && make install clean
fi




if [ -e "$PREFIX/lib/libogg.a" ]
then
	echo "libogg is already installed"
else
	clear
	echo -------------------------
	echo "Downloading and building libogg"
	echo -------------------------
	cd $PREFIX/dist
	${WGET} http://downloads.xiph.org/releases/ogg/libogg-1.2.0.tar.gz
	tar -xvf libogg-1.2.0.tar.gz
	cd libogg-1.2.0
	./configure --enable-shared=no --prefix=$PREFIX
	make && make install clean
	cd $PREFIX/dist
fi



if [ -e "$PREFIX/lib/libvorbis.a" ]
then
	echo "libvorbis is already installed"
else
	clear
	echo -------------------------
	echo "Downloading and building libvorbis"
	echo -------------------------
	${WGET} http://downloads.xiph.org/releases/vorbis/libvorbis-1.3.1.tar.gz
	tar -xvf libvorbis-1.3.1.tar.gz
	cd libvorbis-1.3.1
	./configure --enable-shared=no --prefix=$PREFIX
	make && make install clean
fi



if [ -e "$PREFIX/lib/libtheora.a" ]
then
	echo "libvorbis is already installed"
else
	clear
	echo -------------------------
	echo "Downloading and building libtheora"
	echo -------------------------
	cd $PREFIX/dist
	${WGET} http://downloads.xiph.org/releases/theora/libtheora-1.1.1.tar.bz2
	tar -xvf libtheora-1.1.1.tar.bz2
	cd libtheora-1.1.1
	./configure --enable-shared=no --prefix=$PREFIX --with-ogg=$PREFIX --with-vorbis=$PREFIX
	make && make install clean
	cd $PREFIX/dist
fi



if [ -e "$PREFIX/lib/libx264.a" ]
then
	echo "x264 is already installed"
else
	clear
	echo -------------------------
	echo "Downloading and building x264"
	echo -------------------------
	${WGET} ftp://ftp.videolan.org/pub/videolan/x264/snapshots/x264-snapshot-20100922-2245.tar.bz2
	tar -xvf x264-snapshot-20100922-2245.tar.bz2
	cd x264-snapshot-20100922-2245
	./configure --prefix=$PREFIX
	make && make install clean
fi


if [ -e "$PREFIX/lib/libxvidcore.a" ]
then
	echo "xvidcore is already installed"
else
	clear
	echo -------------------------
	echo "Downloading and building xvidcore"
	echo -------------------------
	cd $PREFIX/dist
	${WGET} http://downloads.xvid.org/downloads/xvidcore-1.2.2.tar.gz
	tar -xvf xvidcore-1.2.2.tar.gz
	cd xvidcore/build/generic
	./configure --prefix=$PREFIX
	make && make install clean
fi




#test if cmake is installed.
type -P cmake &>/dev/null || { 
	clear
	echo -------------------------
	echo "Downloading and building CMake"
	echo -------------------------
	cd $PREFIX/dist
	${WGET} http://www.cmake.org/files/v2.8/cmake-2.8.2.tar.gz
	tar -xvf cmake-2.8.2.tar.gz
	cd cmake-2.8.2
	./configure --prefix=$PREFIX
	make && make install clean
}


# HACK ALERT!
# OpenCV ships with a bunch of libraries (Jasper, Jpeg, PNG, TIFF) in a folder called "3rdParty"
# I'd like to be able to compile my own versions of these libraries and have OpenCV use them
# instead of the who-knows-how-outdated versions that are in there. 
# The FFMPEG Makefile needs CFLAGS, CPPFLAGS, LDFLAGS set so that it knows where to find 
# stuff like x264, lame, vorbis, etc. 
# But if we install Jasper, Jpeg, PNG, TIFF separately and then try to install OpenCV, we get an 
# error.  Sometimes with JPEG, sometimes with TIFF, depending on the version of OpenCV.
# I guess it confuses the headers that have already been installed with the ones that it is trying to compile?
# I can't seem to stop OpenCV from making them, even with OPENCV_BUILD_3RDPARTY_LIBS=FALSE
# So, there are 2 options: 
# 1. We can compile Jasper, JPEG, TIFF, and then unset these variables before compiling OpenCV so that
# 	it doesn't get confused.  Problem:  OpenCV is going to link against its own "3rdParty" versions of 
#	the libraries no matter what, and it *might* cause problems later on to use a different version
# 2. We can just suck it up and use the versions that come with OpenCV.  To do this, we need to
#	not have these varialbes (which are required for FFMPEG to find x264, faac, etc.) set.

if [ -e "$PREFIX/lib/libopencv_core.a" ]
then
	echo "OpenCV is already installed"
else
	export CFLAGS=""
	export CPPFLAGS=""
	export LDFLAGS=""
	clear
	echo -------------------------
	echo "Checking out and building OpenCV"
	echo -------------------------
	cd $PREFIX/dist;
	svn -r 4374 co https://code.ros.org/svn/opencv/trunk/opencv opencv2.2
	cd opencv2.2
	cmake -G "Unix Makefiles" -D CMAKE_OSX_ARCHITECTURES="i386;x86_64" -D BUILD_EXAMPLES=OFF -D BUILD_NEW_PYTHON_SUPPORT=OFF -D BUILD_TESTS=OFF -D BUILD_SHARED_LIBS=OFF -D CMAKE_BUILD_TYPE=Debug -D CMAKE_INSTALL_PREFIX=$PREFIX .
	make && make install
	export CFLAGS="-I$PREFIX/include"
	export CPPFLAGS="-I$PREFIX/include"
	export LDFLAGS="-L$PREFIX/lib"
fi


if [ -e "$PREFIX/include/gavl/gavl.h" ]
then
	echo "GAVL is already installed"
else
	clear
	echo -------------------------
	echo "Downloading and building gavl-1.1.1"
	echo -------------------------
	cd $PREFIX/dist
	${WGET} http://downloads.sourceforge.net/project/gmerlin/gavl/1.1.1/gavl-1.1.1.tar.gz
	tar -xvf gavl-1.1.1.tar.gz
	cd $PREFIX/dist/gavl-1.1.1
	./configure --prefix=$PREFIX --without-doxygen --disable-libpng
	make
	make install
fi



if [ -e "$PREFIX/include/frei0r.h" ]
then
	echo "Frei0r is already installed"
else
	clear
	echo -------------------------
	echo "Downloading and building frei0r-plugins-1.2"
	echo -------------------------
	export LDFLAGS="-L$PREFIX/lib -framework vecLib"
	cd $PREFIX/dist
	${WGET} http://www.piksel.no/frei0r/releases/frei0r-plugins-1.2.tar.gz
	tar -xvf frei0r-plugins-1.2.tar.gz
	cd $PREFIX/dist/frei0r-plugins-1.2
	./configure --prefix=$PREFIX
	# the 'make' will fail because of a problem with the documentation.  But the build did succeed
	make --keep-going --ignore-errors
	make install
fi



clear
echo -------------------------
echo "Checking out and building FFMPEG"
echo -------------------------
cd $PREFIX/dist
svn co -r 26206 svn://svn.ffmpeg.org/ffmpeg/trunk ffmpeg 
cd $PREFIX/dist/ffmpeg
./configure --prefix=$PREFIX --enable-frei0r --disable-asm --disable-stripping --enable-debug --enable-gpl --disable-doc --enable-libmp3lame --enable-libx264 --enable-nonfree --enable-libvorbis --enable-libxvid --enable-version3 --enable-pthreads --enable-libfaac --enable-libopencore-amrnb --enable-libopencore-amrwb --enable-libtheora --enable-libxvid --enable-x11grab
make && make install


#ftp://ftp.remotesensing.org/pub/libtiff/tiff-3.9.4.tar.gz
#ftp://ftp.simplesystems.org/pub/libpng/png/src/libpng-1.4.4.tar.gz
#http://www.ijg.org/files/jpegsrc.v8b.tar.gz
#http://www.ece.uvic.ca/~mdadams/jasper/software/jasper-1.900.1.zip



clear
echo -------------------------
echo "Done!"
echo -------------------------
exit 0
